rbx_reflection_database/src/lib.rs

```diff
@@ -1,15 +1,144 @@
+//! only loading the locally stored database or only the bundled one
+//! respectively.
+//!
+//! ## Local Details
+//!
+//! This crate will load a reflection database from the file system if one
+//! exists in the default location. This location varies upon the OS and is
+//! specified here:
+//!
+//! | OS      | Location                                                            |
+//! |:--------|:--------------------------------------------------------------------|
+//! | Windows | `%localappdata%/.rbxreflection/database.msgpack`                    |
+//! | MacOS   | `$HOME/Library/Application Support/.rbxreflection/database.msgpack` |
+//! | Linux   | `$HOME/.rbxreflection/database.msgpack`                             |
+//!
+//! Additionally, a location override may be specified via the `RBX_DATABASE`
+//! environment variable. The `RBX_DATABASE` variable points to the override
+//! `database.msgpack` file, _not_ to an override `.rbxreflection` directory.
+//!
+//! Both the default `database.msgpack` files and any files pointed to by
+//! `RBX_DATABASE` must be valid MessagePack serializations of a
+//! [`ReflectionDatabase`] if they're present.
+mod error;
+
 use rbx_reflection::ReflectionDatabase;
 
+use std::{env, fs, path::PathBuf, sync::OnceLock};
+
+pub use error::Error;
+
+/// An alias to avoid overly verbose types.
+type ResultOption<T> = Result<Option<T>, Error>;
+
 static ENCODED_DATABASE: &[u8] = include_bytes!("../database.msgpack");
 
+/// The name of an environment variable that may be used to specify
+/// the location of a reflection database to use. The expected format of
+/// a file at this point is MessagePack.
+pub const OVERRIDE_PATH_VAR: &str = "RBX_DATABASE";
+
+/// The name of the directory used for the local location for a reflection
+/// database. The directory will be placed inside the current user's
+/// local data folder on MacOS and Windows and inside
+/// the home directory on Linux.
+pub const LOCAL_DIR_NAME: &str = ".rbxreflection";
+
 lazy_static::lazy_static! {
-    static ref DATABASE: ReflectionDatabase<'static> = {
+    static ref BUNDLED_DATABASE: ReflectionDatabase<'static> = {
+        log::debug!("Loading bundled reflection database");
         rmp_serde::decode::from_slice(ENCODED_DATABASE).unwrap_or_else(|e| panic!("could not decode reflection database because: {}", e))
     };
 }
 
-pub fn get() -> &'static ReflectionDatabase<'static> {
-    &DATABASE
+static LOCAL_DATABASE: OnceLock<ResultOption<ReflectionDatabase<'static>>> = OnceLock::new();
+
+/// Returns a populated [`ReflectionDatabase`]. This will attempt to load one locally and
+/// if one can't be found, it will return one that is bundled with this crate.
+///
+/// ## Errors
+///
+/// Errors if a locally stored [`ReflectionDatabase`] could not be read
+/// or is invalid MessagePack.
+pub fn get() -> Result<&'static ReflectionDatabase<'static>, Error> {
+    Ok(get_local()?.unwrap_or(&BUNDLED_DATABASE))
+}
```

Comment:
I'm not totally sure if breaking this interface is the right call. The way I'm thinking about it is: if `get_local` fails, then probably 99% of the time, users of this function will immediately want to call `get_bundled`, so why not just have this function do that for them? That way it does the right thing most of the time, and we can avoid breakage.

On the other hand, then it's not clear that this function can fail, and removes the ability to e.g. print a warning if the local database is corrupt while using this function.